mri/reconstructors/base.py

```diff
@@ -0,0 +1,155 @@
+
+# System import
+import warnings
+
+# Package import
+from ..operators.linear.wavelet import WaveletUD2, WaveletN
+from ..optimizers import pogm, condatvu, fista
+from ..optimizers.utils.cost import GenericCost
+
+# Third party import
+from modopt.opt.linear import Identity
+
+
+class ReconstructorBase(object):
+    """ This is the base reconstructor class for reconstruction.
+    This class holds some parameters that are common for all MR Image
+    reconstructors
+    For the Analysis case, finds the solution  for x of:
+        (1/2) * ||F x - y||^2_2 + mu * H (W x)
+
+    For the Synthesis case, finds the solution of:
+        (1/2) * ||F Wt alpha - y||^2_2 + mu * H(alpha)
+    Parameters
+    ----------
+    fourier_op: object of class FFT, NonCartesianFFT or Stacked3DNFFT in
+                mri.operators
+        Defines the fourier operator F.
+    linear_op: object
+        Defines the linear sparsifying operator W. This must operate on x and
+        have 2 functions, op(x) and adj_op(coeff) which implements the
+        operator and adjoint operator. For wavelets, this can be object of
+        class WaveletN or WaveletUD2 from mri.operators .
+    regularizer_op: operator, (optional default None)
+        Defines the regularization operator for the regularization function H.
+        If None, the  regularization chosen is Identity and the optimization
+        turns to gradient descent.
+    gradient_formulation: str between 'analysis' or 'synthesis',
+        default 'synthesis'
+        defines the formulation of the image model which defines the gradient.
+    lips_calc_max_iter: int
+        Defines the maximum number of iterations to calculate the lipchitz
+        constant
+    num_check_lips: int
+        Number of iterations to check if the lipchitz constant is correct
+    lipschitz_cst: int, default None
+        The user specified lipschitz constant. If this is not specified,
+        it is calculated using PowerMethod
+    verbose: int
+        Verbosity level.
+            1 => Print basic debug information
+            5 => Print all initialization information
+            20 => Calculate cost at the end of each iteration.
+                NOTE : This is computationally intensive.
+            30 => Print the debug information of operators if defined by class
+    init_gradient_op: bool, default True
+        This parameter controls whether the gradient operator must be
+        initialized right now.
+        If set to false, the user needs to call initialize_gradient_op to
+        initialize the gradient at right time before reconstruction
+    Note:
+    -----
+    The user is expected to specify the either prox_op or mu to obtain
+    reconstructions, else the above equations lose the regularization terms
+    resulting in inverse transform as solution.
+    """
+
+    def __init__(self, fourier_op, linear_op, regularizer_op,
+                 gradient_formulation, grad_class, lips_calc_max_iter,
```

Comment:
What is grad class (it is not Documented)

Reply:
@zaccharieramzi @chaithyagr , isn't it kind of weird to give to the Base reconstructor the parameter `grad_class`, first this is redundant with respect to the `gradient_formulation`. Basically if the `gradient_formulation` is synthesis then the `grad_class` should be `GradientSynthesis` otherwise it should be `GradientAnalysis`.
Or maybe since tis is the Base reconstructor we should consider giving the gradient as parameter in the initialization

Reply:
No.
If we have Self Calibrating reconstruction, it would be GradientSelfCalibratingAnalysis or GradientSelfCalibratingSynthesis

Reply:
So I guess that the best thing to do would be to give the gradient as a parameter of the Base Reconstructor. 
This solution of providing the Gradient Class and the gradient argument is not clean at all

Reply:
I will resolve this as we are addressing the same issue at https://github.com/CEA-COSMIC/pysap-mri/pull/57#discussion_r350738454